// containers/beers/index.js
import React, { Component } from 'react'
import { connect } from 'react-redux'
import { Link } from 'react-router-dom'

import { fetchBeers, loadMoreBeers } from 'actions'
import { getBeers } from 'selectors'

class Beers extends Component {

    componentDidMount () {
        this.props.fetchBeers()
        window.addEventListener('scroll', this.infiniteScroll.bind(this))
    }

    infiniteScroll() {
        const windowHeight = 'innerHeight' in window ? window.innerHeight : document.documentElement.offsetHeight;
        const body = document.body;
        const html = document.documentElement;
        const docHeight = Math.max(body.scrollHeight, body.offsetHeight, html.clientHeight, html.scrollHeight);
        const windowBottom = windowHeight + window.pageYOffset;

        if (windowBottom >= docHeight) {
            console.log('you reached the bottom')
            this.props.loadMoreBeers()
        }
    }

    renderBeer (beer, index) {
        return (
            <div className='col-md-4' key={index}>
                <div className='thumbnail'>
                    <img className='img-thumbnail' 
                    src={beer.image_url} 
                    alt={beer.name} />
                    <div className='caption'>
                        <h4>{beer.name}</h4>
                        <p>{beer.tagline}</p>
                    </div>
                </div>
            </div>
        )
    }

    render () {
        const {beers} = this.props
        return (
            <div>
                <div className='row'>
                    {beers.map((beer, index) => this.renderBeer(beer, index))}
                </div>
            </div>
        )
    
    }
}

const mapStateToProps = state => ({
    beers: getBeers(state)
})

const mapDispatchToProps = {
    fetchBeers,
    loadMoreBeers
}

export default connect(mapStateToProps, mapDispatchToProps)(Beers)
